src/index/blockfilterindex.cpp

```diff
@@ -0,0 +1,272 @@
+        DBHeightKey(int height_in) : height(height_in) {}
+
+        template<typename Stream>
+        void Serialize(Stream& s) const
+        {
+            ser_writedata8(s, DB_BLOCK_HEIGHT);
+            ser_writedata32be(s, height);
+        }
+
+        template<typename Stream>
+        void Unserialize(Stream& s)
+        {
+            char prefix = ser_readdata8(s);
+            if (prefix != DB_BLOCK_HEIGHT) {
+                throw std::ios_base::failure("Invalid format for DB key");
+            }
+            height = ser_readdata32be(s);
+        }
+    };
+
+    struct DBHashKey
+    {
+        uint256 hash;
+
+        DBHashKey(const uint256& hash_in) : hash(hash_in) {}
+
+        ADD_SERIALIZE_METHODS;
+
+        template <typename Stream, typename Operation>
+        inline void SerializationOp(Stream& s, Operation ser_action) {
+            char prefix = DB_BLOCK_HASH;
+            READWRITE(prefix);
+            if (prefix != DB_BLOCK_HASH) {
+                throw std::ios_base::failure("Invalid format for DB key");
+            }
+
+            READWRITE(hash);
+        }
+    };
+};
+
+BlockFilterIndex::BlockFilterIndex(BlockFilterType filter_type,
+                                   size_t n_cache_size, bool f_memory, bool f_wipe)
+    : m_filter_type(filter_type)
+{
+    const std::string& filter_name = BlockFilterTypeName(filter_type);
+    if (filter_name == "") throw std::invalid_argument("unknown filter_type");
+
+    fs::path path = GetDataDir() / "indexes" / "blockfilter" / filter_name;
+    fs::create_directories(path);
+
+    m_name = filter_name + " block filter index";
+    m_db = MakeUnique<BaseIndex::DB>(path / "db", n_cache_size, f_memory, f_wipe);
+    m_filter_fileseq = MakeUnique<FlatFileSeq>(std::move(path), "fltr", FILE_CHUNK_SIZE);
+}
+
+bool BlockFilterIndex::Init()
+{
+    if (!m_db->Read(DB_FILTER_POS, m_next_filter_pos)) {
+        m_next_filter_pos.nFile = 0;
+        m_next_filter_pos.nPos = 0;
+    }
+    return BaseIndex::Init();
+}
+
+bool BlockFilterIndex::Commit(CDBBatch& batch)
+{
+    const FlatFilePos& pos = m_next_filter_pos;
+
+    // Flush current filter file to disk.
+    CAutoFile file(m_filter_fileseq->Open(pos), SER_DISK, CLIENT_VERSION);
+    if (file.IsNull()) {
+        return error("%s: Failed to open filter file %d", __func__, pos.nFile);
+    }
+    if (!FileCommit(file.Get())) {
+        return error("%s: Failed to commit filter file %d", __func__, pos.nFile);
+    }
+
+    batch.Write(DB_FILTER_POS, pos);
+    return BaseIndex::Commit(batch);
+}
+
+size_t BlockFilterIndex::WriteFilterToDisk(FlatFilePos& pos, const BlockFilter& filter)
+{
+    assert(filter.GetFilterType() == GetFilterType());
+
+    size_t data_size =
+        GetSerializeSize(filter.GetBlockHash(), CLIENT_VERSION) +
+        GetSerializeSize(filter.GetEncodedFilter(), CLIENT_VERSION);
+
+    // If writing the filter would overflow the file, flush and move to the next one.
```

Comment:
I agree, but the logic for block/undo files is weird because they are synchronized (like the undo file numbers increment when the block file numbers do, not at a size limit).

Would be a good thing to look at refactoring after this is merged maybe.